test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc

```diff
@@ -0,0 +1,187 @@
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake-step-controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepControllerConfigFactory, ValidateConfigWithValidConfigReturnsOk) {
+  FakeStepControllerConfig config;
+  config.set_fixed_rps_value(1);
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake-step-controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepController, GetCurrentCommandLineOptionsReturnsRpsFromConfig) {
+  FakeStepControllerConfig config;
+  config.set_fixed_rps_value(5678);
+  FakeStepController step_controller(config, CommandLineOptions());
+  EXPECT_EQ(step_controller.GetCurrentCommandLineOptions().value().requests_per_second().value(),
+            5678);
+}
+
+TEST(FakeStepController, IsConvergedInitiallyReturnsFalse) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsFalseAfterBenchmarkResultWithoutPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsTrueAfterBenchmarkResultWithPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  MetricEvaluation* evaluation = benchmark_result.mutable_metric_evaluations()->Add();
+  evaluation->set_threshold_score(1.0);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_TRUE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsDoomedReturnsFalseAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  EXPECT_FALSE(step_controller.IsDoomed(doomed_reason));
+}
+
+TEST(FakeStepController, IsDoomedDoesNotWriteDoomedReasonAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason = "untouched";
+  ASSERT_FALSE(step_controller.IsDoomed(doomed_reason));
+  EXPECT_EQ(doomed_reason, "untouched");
```

Comment:
Why do we expect these to be equal if it does not write doomed reason after successful benchmark result? Shouldn't it be empty or something?

Reply:
The local variable doomed_reason starts out with the value "untouched" (line 141) before we call step_controller.IsDoomed(doomed_reason). So the value "untouched" should still be intact instead of getting overwritten somewhere inside IsDoomed.